Домашняя работа 5

khitao opened the pull request:
Запустил локальную БД H2. Описал структуру сущностей City, CityWeather, WeatherType в скриптах Liquibase (schema.sql).

Реализовал CRUD-операции в соответствующих репозиториях для каждой модели, используя JDBC и Hibernate.

Обновил openapi.yml

khitao committed:
Домашняя работа 5: Реализация работы с БД H2

Запустил локальную БД H2. Описал структуру сущностей City, CityWeather, WeatherType в скриптах Liquibase (schema.sql).

Реализовал CRUD-операции в соответствующих репозиториях для каждой модели, используя JDBC и Hibernate.

Обновил openapi.yml

madoed commented on SpringBootApp/src/main/java/ru/khodov/springbootapp/model/City.java:
```diff
@@ -0,0 +1,20 @@
+@Entity
+@Table(name = "city")
+@Data
+public class City {
```
Для Entity должен присутствовать  конструктор без аргументов

madoed commented on SpringBootApp/src/main/java/ru/khodov/springbootapp/model/CityWeather.java:
```diff
@@ -0,0 +1,38 @@
+@Entity
+@Table(name = "city_weather")
+@Data
+public class CityWeather {
```
Для Entity должен присутствовать конструктор без аргументов

madoed commented on SpringBootApp/src/main/java/ru/khodov/springbootapp/model/WeatherType.java:
```diff
@@ -0,0 +1,22 @@
+@Entity
+@Table(name = "weather_type")
+@Data
+public class WeatherType {
```
Для Entity должен присутствовать конструктор без аргументов

madoed commented on SpringBootApp/src/main/java/ru/khodov/springbootapp/model/WeatherType.java:
```diff
@@ -0,0 +1,22 @@
+
+
+
+
```
Зачем столько пустых строк?

madoed commented on SpringBootApp/src/main/java/ru/khodov/springbootapp/repositories/CityJdbcRepository.java:
```diff
@@ -0,0 +1,39 @@
+    private final JdbcTemplate jdbcTemplate;
+
+    public List<City> getAllCities() {
+        return jdbcTemplate.query("SELECT * FROM city", new BeanPropertyRowMapper<>(City.class));
```
BeanPropertyRowMapper необходимо использовать с осторожностью. Но в случае 2ух полей подойдет.

Но я все же предложил написать свою реализацию RowMapper под City

madoed commented on SpringBootApp/src/main/resources/schema.sql:
```diff
@@ -0,0 +1,44 @@
+    UNIQUE (name)
+);
+
+
+INSERT INTO city (name)
+VALUES ('Moscow'),
+       ('Saint Petersburg'),
+       ('Novosibirsk');
+
+CREATE TABLE weather_type
+(
+    id   BIGINT AUTO_INCREMENT PRIMARY KEY,
+    type VARCHAR(255) NOT NULL,
+    UNIQUE (type)
+);
+
+
+INSERT INTO weather_type (type)
+VALUES ('Sunny'),
+       ('Rainy'),
+       ('Snowy');
+
+
+CREATE TABLE city_weather
+(
+    id              BIGINT AUTO_INCREMENT PRIMARY KEY,
+    city_id         BIGINT           NOT NULL,
+    weather_type_id BIGINT           NOT NULL,
+    temperature     FLOAT NOT NULL,
+    date_time       TIMESTAMP     NOT NULL,
+    FOREIGN KEY (city_id) REFERENCES city (id) ON DELETE CASCADE,
+    FOREIGN KEY (weather_type_id) REFERENCES weather_type (id)
+);
+
+
+INSERT INTO city_weather (city_id, weather_type_id, temperature, date_time)
+VALUES (1, 1, 20.5, '2023-10-14 12:00:00'),
+       (2, 2, 15.3, '2023-10-14 12:00:00'),
+       (1, 2, 15.3, '2023-10-14 12:00:00'),
+       (3, 3, -5.0, '2023-10-14 12:00:00');
```
В задании сказано использовать инструмент миграции Liquibase

Необходимо переделать

madoed commented on SpringBootApp/build.gradle:
```diff
@@ -39,3 +42,7 @@ dependencies {
     useJUnitPlatform()
 }
+
+compileJava {
+    options.compilerArgs << "-Xlint:deprecation"
+}
```
Зачем?

khitao replied:
IDEA рекомендовала добавить, так как в проекте был устаревший метод. Забыл убрать после внесения изменений

madoed commented on SpringBootApp/src/main/java/ru/khodov/springbootapp/model/CityWeather.java:
```diff
@@ -0,0 +1,38 @@
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    @ManyToOne()
```
Пустые скобки не нужны

madoed commented on SpringBootApp/src/main/java/ru/khodov/springbootapp/repositories/CityWeatherJdbcRepository.java:
```diff
@@ -0,0 +1,52 @@
+    private final JdbcTemplate jdbcTemplate;
+
+    public List<CityWeather> getAllCityWeather() {
+        String sql = "SELECT cw.id, cw.temperature, cw.date_time, c.id as city_id, c.name as city_name, wt.id as weather_type_id, wt.type as weather_type " +
+                "FROM city_weather cw " +
+                "LEFT JOIN city c ON cw.city_id = c.id " +
+                "LEFT JOIN weather_type wt ON cw.weather_type_id = wt.id";
```
Можно использовать text blocks https://www.baeldung.com/java-text-blocks

khitao committed:
Исправления домашняя работа 5

Добавил конструкторы без аргументов для каждого Entity

Реализовал свои RowMapper для City и WeatherType

Использовал инструмент миграции Liquibase

Применил text blocks для улучшения читаемости кода
